### PR TITLE
Support symbolizing enums

### LINKED ISSUE #2772: Support symbolizing enum values
### Is your feature request related to a problem? Please describe.

Consider:
```
$ sudo bpftrace -e 'tracepoint:skb:kfree_skb { print(args->reason) }'
Attaching 1 probe...
50
50
```

bpftrace knows the type of `args->reason`:
```
$ sudo bpftrace -lv tracepoint:skb:kfree_skb
tracepoint:skb:kfree_skb
    void * skbaddr
    void * location
    unsigned short protocol
    enum skb_drop_reason reason
```

as well as the names of the values:
```
 $ sudo bpftrace -lv 'enum skb_drop_reason' | grep 50
        SKB_DROP_REASON_UNHANDLED_PROTO = 50,
```

Would be nice if bpftrace could be printing "SKB_DROP_REASON_UNHANDLED_PROTO" instead of integer values.

### Describe the solution you'd like

Ideally we just do it and tell users who want the integer values to cast the enum to an int. But that might not be backwards compatible.

### Describe alternative solutions or features you've considered

 Could reuse `str()` but overload on identifiers. That's kinda ugly though IMO. Could create a new helper. Also prefer to avoid that. Wonder how many people are depending on enums being printed as an integer. Seems like it's just an annoyance for everyone.
Maybe we can have an environ variable to control this? Not sure.

### PR DESCRIPTION
**Pull request by danobi**
This PR allows the user to symbolize enums in userspace by
using `%s` printf() specifier. This is a fully backwards compatible
change, as trying to use `%s` on an enum was a compile error
before.

Example:
```
# bpftrace -e 'tracepoint:skb:kfree_skb { $r = args->reason; printf("%d, %s\n", $r, $r); exit() }'
Attaching 1 probe...
2, SKB_DROP_REASON_NOT_SPECIFIED
```

Note this PR does not support casting integers to enums. While
this is technically possible (and described in the linked ticket), it's
probably not necessary as users can simply perform arithmetic
operations on an enum and get the same result. There's currently
a limitation but it's being fixed in #3517.

This is easier to review per-commit.

This closes #2772.

##### Checklist

- [x] Language changes are updated in `man/adoc/bpftrace.adoc`
- [x] User-visible and non-trivial changes updated in `CHANGELOG.md`
- [x] The new behaviour is covered by tests

### CODE REVIEW

**Review by ajor — CHANGES_REQUESTED** [review]
Going back to our discussions in the issue: I'm not sure that this PR solves enough of the problem to be useful, and by defining enums as integers, it risks locking us into behaviour that we won't be able to change to fix it in the future.

We need to be able to symbolicate enums everywhere that they can appear:
- as a field in a struct: `{ .myenum = ABC }`
- in tuples: `(SKB_CONSUMED, 123)`
- as map keys or values: `@[SKB_CONSUMED] = ABC`

**Comment by danobi** [comment]
> Going back to our discussions in the issue: I'm not sure that this PR solves enough of the problem to be useful,

I think it's a good start. A few users have asked me for this already. Let me follow up and collect their feedback on this approach.

> and by defining enums as integers, it risks locking us into behaviour that we won't be able to change to fix it in the future.

I think it might be too late for enums as they exist today. Plenty of people are using today's enums as ints. But the good news is I haven't seen it be a problem yet in the ~5 years they've been in: https://github.com/bpftrace/bpftrace/commit/e4cb6ce47018bf857ed0dbd0f0e01ce4e9cb63c9

> We need to be able to symbolicate enums everywhere that they can appear:

For sure, agreed. But I think we need to start somewhere and printf() is the highest impact for now. We'll need the metadata collected in commits 1-6 to keep building on top.

**Comment by danobi** [comment]
@ajor and I chatted offline. Decision is that this approach is ok. We'll decide how to extend enum symbolication to struct fields, tuples, and map key/value next (or even at some other later point). Gut feeling is map key is next most useful.

It won't be hard to do when the infra in this PR is landed. Just a question of how much backwards compat we want. For example, changing human readable output is probably fine. But we probably want to keep JSON output as-is so tooling doesn't silently break.

File: src/format_string.cpp [hunk 1]
```diff
@@ -39,6 +41,37 @@ ArgumentType get_expected_argument_type(const std::string &fmt)
   return ArgumentType::UNKNOWN;
 }
 
+// Returns a vector of (token, type) tuples given a format string
+std::vector<std::tuple<std::string, Type>> get_token_types(
```
Comment (jordalgo) [inline src/format_string.cpp:45]: nit: Maybe a few unit tests for this would be good.

File: src/format_string.h [hunk 2]
```diff
@@ -34,8 +37,12 @@ class FormatString {
    */
   FormatString() = default;
 
-  FormatString(const char *s) : fmt_(s){};
-  FormatString(std::string &s) : fmt_(s){};
+  FormatString(const char *s) : fmt_(s)
+  {
```
Comment (jordalgo) [inline src/format_string.h:41]: why this change?

File: src/format_string.h [hunk 3]
```diff
@@ -82,14 +89,15 @@ class FormatString {
...
 
   friend class cereal::access;
 
   template <typename Archive>
   void serialize(Archive &ar)
   {
-    // NOTE: parts_ is not constructed until first use, so no point in
-    // serializing it
+    // NOTE: parts_, expected_types_, and tokens_ are not constructed until
+    // first use, so no point in serializing it
```
Comment (jordalgo) [inline src/format_string.h:100]: "so no point in serializing them"

File: tests/runtime/call [hunk 4]
```diff
@@ -30,6 +30,28 @@ NAME printf_char
...
+
+NAME printf_enum_symbolize_map
+PROG enum Foo { ONE = 1, TWO = 2, OTHER = 99999 }; BEGIN { @one = ONE; @two = TWO; @other = OTHER; printf("%d %s %d %s %d %s\n", @one, @one, @two, @two, @other, @other); exit() }
+EXPECT 1 ONE 2 TWO 99999 OTHER
+
+NAME printf_enum_symbolize_width
+PROG enum Foo { A, B, C, }; BEGIN { printf("%-5s %5s %s\n", A, B, C); exit() }
+EXPECT A         B C
+
+NAME printf_enum_symbolize_tracepoint
```
Comment (jordalgo) [inline tests/runtime/call:49]: should we add a test for trying to symbolize an int that isn't an enum?

File: man/adoc/bpftrace.adoc [hunk 5]
```diff
@@ -1922,7 +1922,26 @@ The non-standard ones can be found in the table below:
...
+  CUSTOM_ENUM = 3,
+};
+
+tracepoint:skb:kfree_skb {
+  $r = args->reason;
+  printf("%d, %s, %s\n", $r, $r, CUSTOM_ENUM);
+}
+----
+
+may yield:
```
Comment (jordalgo) [inline man/adoc/bpftrace.adoc:1938]: This feels confusing. Why not just use a `BEGIN` probe so you can say what it definitely will print out.

**Comment by jordalgo** [comment]
> But we probably want to keep JSON output as-is so tooling doesn't silently break.

Would be nice if we could provide a way to print the symbolized enums in JSON output though - env variable?

File: src/format_string.h (hunk 2 above)
Comment (danobi) [reply src/format_string.h:41]: dunno. clang-format really wants to do this and it won't let me leave it unchanged :P

File: tests/runtime/call (hunk 4 above)
Comment (danobi) [reply tests/runtime/call:49]: Yeah makes sense. I'll add it to semantic analyser tests cuz it should be a compile failure

File: man/adoc/bpftrace.adoc (hunk 5 above)
Comment (danobi) [reply man/adoc/bpftrace.adoc:1938]: Yeah, agreed. Simplified.

File: src/format_string.cpp (hunk 1 above)
Comment (danobi) [reply src/format_string.cpp:45]: Took a look. Could probably be done, but would need to pull this out of anon namespace. We also already have quite a bit of coverage in semantic analyser tests. See `semantic_analyser.printf_*` set of tests. So I think coverage is likely sufficient here. WDYT?

**Comment by danobi** [comment]
> > But we probably want to keep JSON output as-is so tooling doesn't silently break.
> 
> Would be nice if we could provide a way to print the symbolized enums in JSON output though - env variable?

That was referring to changing the output for map keys/values, tuples, and structs. I think changing the human output is fine but JSON is dicey. As folks may be parsing it in prod already.

For regular printf() (added in this PR), JSON output changes "correctly":

```
$ sudo ./build/src/bpftrace -e 'BEGIN { printf("%s\n", SKB_DROP_REASON_NOT_SPECIFIED); exit() }' -f json
{"type": "attached_probes", "data": {"probes": 1}}
{"type": "printf", "data": "SKB_DROP_REASON_NOT_SPECIFIED\n"}
```

**Comment by jordalgo** [comment]
> I think changing the human output is fine but JSON is dicey. As folks may be parsing it in prod already. 

I guess we'll see if we get requests from users to make this configurable.

File: src/bpftrace.h [hunk 6]
```diff
@@ -182,7 +182,10 @@ class BPFtrace {
   BpfBytecode bytecode_;
   StructManager structs;
   std::map<std::string, std::string> macros_;
-  std::map<std::string, uint64_t> enums_;
+  // Map of enum variant_name -> (variant_value, enum_name)
```
Comment (jordalgo) [inline src/bpftrace.h:185]: super nit (just language consistency)
```suggestion
  // Map of enum variant_name to (variant_value, enum_name)
```

File: src/types.h [hunk 7]
```diff
@@ -302,7 +303,7 @@ class SizedType {
 
   const std::string GetName() const
   {
-    assert(IsRecordTy());
+    assert(IsRecordTy() || IsIntTy());
```
Comment (jordalgo) [inline src/types.h:306]: Nit. Does this work?
```suggestion
    assert(IsRecordTy() || IsEnumTy());
```

**Review by jordalgo — APPROVED** [review]
One super nit and one question but otherwise LGTM.

File: src/types.h (hunk 7 above)
Comment (danobi) [reply src/types.h:306]: Ah yeah that's better
